src/fpm/manifest/profiles.f90

```diff
@@ -0,0 +1,732 @@
+
+        character(len=:), allocatable :: flags, c_flags, link_time_flags, key_name, file_name, file_flags
+        type(toml_table), pointer :: files
+        type(toml_key), allocatable :: file_list(:)
+        type(file_scope_flag), allocatable :: file_scope_flags(:)
+        integer :: ikey, ifile, stat
+        logical :: is_valid
+
+        if (size(key_list).ge.1) then
+          do ikey=1,size(key_list)
+            key_name = key_list(ikey)%key
+            if (key_name.eq.'flags') then
+              call get_value(table, 'flags', flags, stat=stat)
+              if (stat /= toml_stat%success) then
+                call syntax_error(error, "flags has to be a key-value pair")
+                return
+              end if
+            else if (key_name.eq.'c_flags') then
+              call get_value(table, 'c_flags', c_flags, stat=stat)
+              if (stat /= toml_stat%success) then
+                call syntax_error(error, "c_flags has to be a key-value pair")
+                return
+              end if
+            else if (key_name.eq.'link_time_flags') then
+              call get_value(table, 'link_time_flags', link_time_flags, stat=stat)
+              if (stat /= toml_stat%success) then
+                call syntax_error(error, "link_time_flags has to be a key-value pair")
+                return
+              end if
+            else if (key_name.eq.'files') then
+              call get_value(table, 'files', files, stat=stat)
+              if (stat /= toml_stat%success) then
+                call syntax_error(error, "files has to be a table")
+                return
+              end if
+              call files%get_keys(file_list)
+              allocate(file_scope_flags(size(file_list)))
+              do ifile=1,size(file_list)
+                file_name = file_list(ifile)%key
+                call get_value(files, file_name, file_flags, stat=stat)
+                if (stat /= toml_stat%success) then
+                  call syntax_error(error, "file scope flags has to be a key-value pair")
+                  return
+                end if
+                associate(cur_file=>file_scope_flags(ifile))
+                  if (.not.(path.eq."")) file_name = join_path(path, file_name)
+                  cur_file%file_name = file_name
+                  cur_file%flags = file_flags
+                end associate
+              end do
+            end if
+          end do
+        end if
+
+        if (.not.allocated(flags)) flags=''
+        if (.not.allocated(c_flags)) c_flags=''
+        if (.not.allocated(link_time_flags)) link_time_flags=''
+
+        if (allocated(file_scope_flags)) then
+          profiles(profindex) = new_profile(profile_name, compiler_name, os_type, &
+                 & flags, c_flags, link_time_flags, file_scope_flags)
+        else
+          profiles(profindex) = new_profile(profile_name, compiler_name, os_type, flags, c_flags, link_time_flags)
+        end if
```

Comment:
Why this code duplication, if file_scope_flags is not allocated the constructor will handle it as not present. Therefore, we only need one call here.